hadoop-hdfs-project/hadoop-hdfs/src/site/markdown/NamenodeFGL.md

```diff
@@ -0,0 +1,201 @@
+
+Locks are to be acquired in this order:
+- Global > DN > Full path
+- Global > DN > Last `INodeFile`
+
+Possible lock combinations are as follows:
+- Global write lock
+- Global read lock > Full path lock
+- Global read lock > DN read/write lock
+- Global read lock > DN read/write lock > Read/Write lock of last `INodeFile`
+- Global read lock > DN read/write lock > Full path lock
+
+### Lock Pools
+
+NN allocates locks as needed to the INodes used by active threads, and deletes them after the locks are no longer in use. Locks for commonly accessed `INode`s like the root are cached.
+
+NN uses an `INodeLockPool` to manage these locks. The lock pool:
+- Returns a closeable lock for an INode based on the lock type,
+- Removes this lock if it is no longer used by any threads.
+
+Similar to `INodeLockPool`, a `DNLockPool` is used to manage the locks for DNs. Unlike `INodeLockPool`, `DNLockPool` keeps all locks in memory due to the comparatively lower number of locks.
+
+### Lock Modes
+
+Operations related to namespace tree have different semantics and may involve the modification or access of different INodes, for example: `getBlockLocation` only accesses the last iNodeFile, `delete` modifies both the parent and the last INode, `mkdir` may modify multiple ancestor INodes.
+
+Four lock modes (plus no locking):
+- LOCK_READ
+  - This lock mode acquires the read locks for all INodes down the path.
+  - Example operations: `getBlockLocation`, `getFileInfo`.
+- LOCK_WRITE
+  - This lock mode acquires the write lock for the last INode and the read locks for all ancestor INodes in the full path.
+  - Example operations: `setPermission`, `setReplication`.
+- LOCK_PARENT
+  - This lock mode acquires the write lock for the last two INodes and the read locks for all remaining ancestor INodes in the full path.
+  - Example operations: `rename`, `delete`, `create` (when the parent directory exists).
+- LOCK_ANCESTOR
+  - This lock mode acquires the write lock for the last existing INode and the read locks for all remaining ancestor INodes in the full path.
+  - Example operations: `mkdir`, `create` (when the parent directory doesn't exist).
+- NONE
+  - This lock mode does not acquire any locks for the given path.
+
+Roadmap
+-------
+
+#### Step 1: Split the global lock into FSLock and BMLock
+
+Split the global lock into two global locks, FSLock and BMLock.
+- FSLock for operations that relate to namespace tree.
+- BMLock for operations related to blocks and/or operations related to DNs.
+- Both FSLock and BMLock for HA related operations.
+After this step, FGL contains global FSLock and global BMLock.
+
+No big logic changes in this step. The original logic with the global lock retains. This step aims to make the lock mode configurable.
+
+JIRA: [HDFS-17384](https://issues.apache.org/jira/browse/HDFS-17384)
+
+#### Step 2: Split the global FSLock
```

Comment:
There are some difficult problems should be fixed in this step. And we have discussed it and got some solutions.

- Quota problem
- Snapshot problem
- StoragePolicy ID problem

I will create some tickets to describe these problems and give some solution in the ticket for these problems in milestone2. I don't know if these problems should be mentioned in this document.  @ferhui @kokonguyen191 What do you think?

Reply:
In my opinion, it is a high level document, no need to add more tech details here. can link the design doc or ticket to the details.

Reply:
We can regard this as a user help document？

Reply:
Got.